solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/EnumerableSet.sol";
+
+import "./interfaces/ILiquidityProtectionStats.sol";
+import "../utility/Utils.sol";
+import "../utility/SafeMath.sol";
+import "../token/interfaces/IDSToken.sol";
+import "../token/interfaces/IERC20Token.sol";
+
+/**
+ * @dev This contract aggregates the statistics of the liquidity protection mechanism.
+ */
+contract LiquidityProtectionStats is ILiquidityProtectionStats, AccessControl, Utils {
+    using SafeMath for uint256;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    bytes32 public constant ROLE_SUPERVISOR = keccak256("ROLE_SUPERVISOR");
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
+    bytes32 public constant ROLE_OWNER = keccak256("ROLE_OWNER");
+
+    mapping(IDSToken => uint256) public totalPoolAmount;
+    mapping(IDSToken => mapping(IERC20Token => uint256)) public totalReserveAmount;
+    mapping(IDSToken => mapping(IERC20Token => mapping(address => uint256))) public totalProviderAmount;
+
+    mapping(address => EnumerableSet.AddressSet) private _providerPools;
+
+    // allows execution by the owner only
+    modifier ownerOnly {
+        require(hasRole(ROLE_OWNER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    // allows execution by the seeder only
+    modifier seederOnly {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    constructor() public {
+        // set up administrative roles
+        _setRoleAdmin(ROLE_SUPERVISOR, ROLE_SUPERVISOR);
+        _setRoleAdmin(ROLE_SEEDER, ROLE_SUPERVISOR);
```

Comment:
I think that the ROLE_OWNER should be the ROLE_SEEDER's admin since ROLE_SUPERVISOR is mostly for backup or emergency intervention

Reply:
So no need for ROLE_SUPERVISOR?

Reply:
ROLE_SUPERVISOR is needed to manage ROLE_OWNER, while ROLE_OWNER will manage all the rest

Reply:
@lbeder : @yudilevi has opted to keep as is.

Reply:
Alrighty